File: go/internal/validation/are_change_files_deleted_test.go
package validation_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/microsoft/beachball/internal/validation"
	"github.com/stretchr/testify/assert"
)

// get default options with a specified root path (fetch disabled)
func getDefaultOptionsWithPath(rootPath string) types.BeachballOptions {
	defaultOptions := types.DefaultOptions()
	defaultOptions.Branch = testutil.DefaultRemoteBranch
	defaultOptions.Fetch = false
	defaultOptions.Path = rootPath

	return defaultOptions
}

// TS: "is false when no change files are deleted"
func TestAreChangeFilesDeleted_FalseWhenNoChangeFilesDeleted(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	// Create a change file on master and push it
	opts := getDefaultOptionsWithPath(repo.RootPath())

	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)
	repo.Push()

	// Checkout a new branch — no deletions
	repo.Checkout("-b", "test-no-delete", testutil.DefaultBranch)

	result := validation.AreChangeFilesDeleted(&opts)
	assert.False(t, result)
}

// TS: "is true when change files are deleted"
func TestAreChangeFilesDeleted_TrueWhenChangeFilesDeleted(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	// Create a change file on master and push it
	opts := getDefaultOptionsWithPath(repo.RootPath())

	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)
	repo.Push()

	// Checkout a new branch and delete the change files using git rm
	repo.Checkout("-b", "test-delete", testutil.DefaultBranch)

	changePath := filepath.Join(repo.RootPath(), opts.ChangeDir)
	repo.Git([]string{"rm", "-r", changePath})
	repo.Git([]string{"commit", "-m", "Delete change files"})

	// Recreate the directory so git can run from it
	os.MkdirAll(changePath, 0o755)

	result := validation.AreChangeFilesDeleted(&opts)
	assert.True(t, result)
}

// TS: "deletes change files when changeDir option is specified"
func TestAreChangeFilesDeleted_WorksWithCustomChangeDir(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()

	opts := getDefaultOptionsWithPath(repo.RootPath())
	opts.ChangeDir = "custom-changes"

	testutil.GenerateChangeFiles(t, []string{"foo"}, &opts, repo)
	repo.Push()

	repo.Checkout("-b", "test-custom-delete", testutil.DefaultBranch)

	changePath := filepath.Join(repo.RootPath(), opts.ChangeDir)
	repo.Git([]string{"rm", "-r", changePath})
	repo.Git([]string{"commit", "-m", "Delete custom change files"})

	// Recreate the directory so git can run from it
	os.MkdirAll(changePath, 0o755)

	result := validation.AreChangeFilesDeleted(&opts)
	assert.True(t, result)
}

File: rust/tests/are_change_files_deleted_test.rs
mod common;

use beachball::types::options::BeachballOptions;
use beachball::validation::are_change_files_deleted::are_change_files_deleted;
use common::change_files::generate_change_files;
use common::repository_factory::RepositoryFactory;
use common::{DEFAULT_BRANCH, make_test_options};

// TS: "is false when no change files are deleted"
#[test]
fn is_false_when_no_change_files_are_deleted() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let options = make_test_options(repo.root_path(), None);
    generate_change_files(&["foo"], &options, &repo);
    repo.push();

    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);

    let options = make_test_options(repo.root_path(), None);
    assert!(!are_change_files_deleted(&options));
}

// TS: "is true when change files are deleted"
#[test]
fn is_true_when_change_files_are_deleted() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let options = make_test_options(repo.root_path(), None);
    generate_change_files(&["foo"], &options, &repo);
    repo.push();

    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);

    // Delete the change files (but keep the directory so git can run from it)
    let change_path = std::path::Path::new(repo.root_path()).join("change");
    for entry in std::fs::read_dir(&change_path).unwrap() {
        let entry = entry.unwrap();
        std::fs::remove_file(entry.path()).unwrap();
    }
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "delete change files"]);

    let options = make_test_options(repo.root_path(), None);
    assert!(are_change_files_deleted(&options));
}

// TS: "deletes change files when changeDir option is specified"
#[test]
fn works_with_custom_change_dir() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();

    let custom_opts = BeachballOptions {
        change_dir: "changeDir".to_string(),
        ..Default::default()
    };

    let options = make_test_options(repo.root_path(), Some(custom_opts.clone()));
    generate_change_files(&["foo"], &options, &repo);
    repo.push();

    repo.checkout(&["-b", "test", DEFAULT_BRANCH]);

    // Delete the change files (but keep the directory so git can run from it)
    let change_path = std::path::Path::new(repo.root_path()).join("changeDir");
    for entry in std::fs::read_dir(&change_path).unwrap() {
        let entry = entry.unwrap();
        std::fs::remove_file(entry.path()).unwrap();
    }
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "delete change files"]);

    let options = make_test_options(repo.root_path(), Some(custom_opts));
    assert!(are_change_files_deleted(&options));
}
